src/moscot/problems/time/_mixins.py

```diff
@@ -427,102 +426,72 @@ def pull(
         df_list = [
-            pd.DataFrame(problem.prior_growth_rates, index=problem.adata.obs.index, columns=cols)
+            pd.DataFrame(problem.prior_growth_rates, index=problem.adata.obs_names, columns=cols)
             for problem in self.problems.values()
         ]
-        tup = list(self)[-1]
-        df_list.append(
-            pd.DataFrame(
-                np.full(
-                    shape=(len(self.problems[tup].adata_tgt.obs), 1),
-                    fill_value=np.nan,
-                ),
-                index=self.problems[tup].adata_tgt.obs.index,
-                columns=cols,
-            )
-        )
-        return pd.concat(df_list, verify_integrity=True)
+        df_1 = pd.concat(df_list, verify_integrity=True)
+        indices_remaining = set(self.adata.obs_names) - set(df_1.index)
+        df_2 = pd.DataFrame(np.nan, index=list(indices_remaining), columns=cols)
+
+        return pd.concat([df_1, df_2], verify_integrity=True)
 
     @property
     def posterior_growth_rates(self: TemporalMixinProtocol[K, B]) -> Optional[pd.DataFrame]:
         """Return the posterior estimate of growth rates of the cells in the source distribution."""
-        # TODO(michalk8): FIXME
         cols = ["posterior_growth_rates"]
         df_list = [
-            pd.DataFrame(problem.posterior_growth_rates, index=problem.adata.obs.index, columns=cols)
+            pd.DataFrame(problem.posterior_growth_rates, index=problem.adata.obs_names, columns=cols)
             for problem in self.problems.values()
         ]
-        tup = list(self)[-1]
-        df_list.append(
-            pd.DataFrame(
-                np.full(
-                    shape=(len(self.problems[tup].adata_tgt.obs), 1),
-                    fill_value=np.nan,
-                ),
-                index=self.problems[tup].adata_tgt.obs.index,
-                columns=cols,
-            )
-        )
-        return pd.concat(df_list, verify_integrity=True)
+        df_1 = pd.concat(df_list, verify_integrity=True)
+        indices_remaining = set(self.adata.obs_names) - set(df_1.index)
+        df_2 = pd.DataFrame(np.nan, index=list(indices_remaining), columns=cols)
+
+        return pd.concat([df_1, df_2], verify_integrity=True)
 
-    # TODO(michalk8): refactor me
     @property
     def cell_costs_source(self: TemporalMixinProtocol[K, B]) -> Optional[pd.DataFrame]:
         """Return the cost of a cell obtained by the potentials of the optimal transport solution."""
-        sol = list(self.problems.values())[0].solution
-        if TYPE_CHECKING:
-            assert isinstance(sol, BaseSolverOutput)
-        if sol.potentials is None:
+        cols = ["cell_cost_source"]
+        sol = list(self.solutions.values())[0]
+        if sol is None or sol.potentials is None:
             return None
+
         df_list = [
             pd.DataFrame(
                 np.array(np.abs(problem.solution.potentials[0])),  # type: ignore[union-attr,index]
```

Comment:
All of this also should apply to `cell_costs_target` below.

Reply:
> Because it's about the absolute value of the derivative, not the direction, right?

Ok, can you then please add to the docs it returns the absolute values of the dual potentials so that it's clear?

Reply:
Actually changed it now to returning not the `abs` value